sycl/doc/ABIPolicyGuide.md

```diff
@@ -0,0 +1,90 @@
+# ABI Policy Guide
+
+## Intro
+
+Application Binary Interface is a contract between binary modules, that defines
+how structures and routines are accessed in machine code. Changing the ABI may
+break backwards compatibility with library for user-developed applications,
+resulting in need to rebuild such applications. The goal of this document is to
+provide guidelines for maintaining the current ABI and mechanisms of notifying
+users about ABI changes.
+
+All ABI changes can be divided into two large groups: breaking and non-breaking.
+A breaking change means that the new binary is incompatible with the previous
+version (i.e. it can not be used as a drop-in replacement). A non-breaking
+change means that the forward compatibility is broken (i.e. the old library
+can be replaced with newer version, but not vice versa).
+
+The following non-exhaustive list contains changes that are considered to be
+breaking:
+
+1. Changing the size of exported symbol (for example, adding new member field
+   to the exported class).
+1. Removing the exported symbol (that includes both changing the signature of
+   exported routine and removing it).
+1. Changing the alignment of exported symbol.
+1. Changing the layout of exported symbol (for example, reordering class field
+   members).
+1. Adding or removing base classes.
+
+Adding a new exported symbol is considered to be non-breaking change.
+
+## ABI Versioning Policy
+
+TBD
+
+## `__SYCL_EXPORT` Macro
+
+The `__SYCL_EXPORT` provides facilities for fine-grained control over exported
+symbols. Mark symbols that are supposed to be accessible by the user and that
+are implemented in the SYCL Runtime library with this macro. Template
+specializations also must be explicitly marked with `__SYCL_EXPORT` macro.
+Symbols not marked `__SYCL_EXPORT` have internal linkage.
+
+A few examples of when it is necessary to mark symbols with the macro:
+
+* The `device` class:
+  - It is defined as API by the SYCL spec.
+  - It is implemented in `device.cpp` file.
+* The `SYCLMemObjT` class:
+  - It is not defined in the SYCL spec, but it is an implementation detail that
+    is accessible by the user (buffer and image inherit from this class).
+  - It has symbols that are implemented in the Runtime library.
+
+When it is not necessary to mark symbols with `__SYCL_EXPORT`:
+* The `buffer` class:
+  - It is defined by the SYCL spec, but it is fully implemented in the headers.
+* The `ProgramManager` class:
+  - It is an implementation detail.
+  - It is not accessed from the header files that are available to users.
+
+## Automated ABI Changes Testing
+
+> The automated tests deal with the most commonly occurring problems, but they
```

Comment:
```suggestion
The automated tests deal with the most commonly occurring problems, but they
```

Reply:
and 2 lines below

Reply:
@kbobrovs I did that on purpose to differentiate this piece from the rest of the text. Any ideas how we can improve its visibility?

Reply:
I see. I just thought this was a typo. No need to change then.